src/vars.c

```diff
@@ -196,16 +196,16 @@ static Obj EvalIsbLVar(Expr expr)
 **
 **  'PrintAssLVar' prints the local variable assignment statement <stat>.
 */
-static void PrintAssLVar(Stat stat)
+static void PrintAssLVar(Obj body, Stat stat)
 {
     Pr( "%2>", 0L, 0L );
     Pr("%H", (Int)NAME_LVAR(READ_STAT(stat, 0)), 0L);
```

Comment:
Actually, of course `NAME_LVAR` is *also* broken, and really requires a pointer to the function object, as only that contains the names of the lvars, in form of the `Obj nams` plist. Well, or the body would have to contain a pointer to the `nams` plist... but I am worried a bit about potential synchronization issues. What if a `T_BODY` is shared by multiple functions (I think this happens with higher order functions: we create a single `T_BODY`, stored in the function expression object; and then each instantiation of the `fexs` into an actual function uses the same existing `T_BODY`.

And in the future, I was envisioning a version of `PrintFunction` which also prints out the values of all `HVARs` used in the function (that would be super handy during debugging). Though I guess that's OK for `PrintFunction`, as it actually can access the function object...

Hrrm to all.

(Also, clearly I overlooked the `READ_STAT` here, and in a few other places... bad)

Reply:
The nams list logically belongs with the body. We have to distinguish between 

- a body (which is essentially  a piece of code with its lexical context (ie the function body that textually surrounds it); 
- a function (or a closure in many languages) which is a GAP object with a local and possibly some higher variables bags associated with it; and
- a stack frame, which is an actual function invocation with values for the parameters, a currently executing statement calling routine and so on.  

the relationships are one-to-many down the page and any piece of information should belong at one of the three levels, but I don't think GAP is entirely clean about this splitting.  It would be good to make all of this tidier.

Reply:
I am not so sure how that could / should be made tidier, though (not in the sense that I dispute the need or the possibility, just that it is not yet clear to me how to get there).

For example, we also need to take kernel function and gac compiled functions (which also may reference HVARs) into account. The former have no T_BODY at all. If we did the minimal thing, and just moved the list of argument+local names (currently named `namesOfArgsAndLocals`, but I am considering to rename it back to `nams`...) into the T_BODY, then all kernel functions would need to get a T_BODY, just to store the argument names. But OK, we can do that.

Next is the number of arguments `nargs` (BTW, I am thinking about renaming that back to `narg`). Should that also be in the T_BODY? I guess not, as there is no need for it there, and we don't distinguish between arguments and locals on the level of the body; at most it is interesting to know which local "slots" are expected to be pre-filled... The same probably holds for `nloc`, the number of locals (indeed, it is not even clear to me whether we even need that -- we should always have `nloc + narg = LEN_LIST(nams)`, no?

For `prof` (profiling information), `envi` (the environment which started this discussion), `locks` (for HPC-GAP), arguably I'd think they belong with the function, not the body... Hm